building/build-debs/homeworld-keysystem/src/keycommon/endpoint/endpoint.go

```diff
@@ -0,0 +1,108 @@
+package endpoint
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+)
+
+type ServerEndpoint struct {
+	rootCAs      *x509.CertPool
+	baseURL      string
+	extraHeaders map[string]string
+	certificates []tls.Certificate
+	timeout      time.Duration
+}
+
+func NewServerEndpoint(url string, authorities *x509.CertPool) (ServerEndpoint, error) {
+	if url == "" {
+		return ServerEndpoint{}, errors.New("empty base URL")
+	}
+	if !strings.HasSuffix(url, "/") {
+		return ServerEndpoint{}, errors.New("base URL must end in a slash")
+	}
+	return ServerEndpoint{rootCAs: authorities, baseURL: url, timeout: time.Second * 30}, nil
+}
+
+func (s ServerEndpoint) BaseURL() string {
+	return s.baseURL
+}
+
+func (s ServerEndpoint) WithHeader(key string, value string) ServerEndpoint {
+	out := s
+	out.extraHeaders = map[string]string{}
+	for k, v := range s.extraHeaders {
+		out.extraHeaders[k] = v
+	}
+	out.extraHeaders[key] = value
+	return out
+}
+
+func (s ServerEndpoint) WithCertificate(cert tls.Certificate) ServerEndpoint {
+	out := s
+	out.certificates = make([]tls.Certificate, len(s.certificates))
+	copy(out.certificates, s.certificates)
+	out.certificates = append(out.certificates, cert)
+	return out
+}
+
+func (s ServerEndpoint) Request(path string, method string, reqbody []byte) ([]byte, error) {
+	if path[0] != '/' {
+		return nil, errors.New("while validating request: path must be absolute")
+	}
+	client := http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs:      s.rootCAs,
+				Certificates: s.certificates,
+			},
+		},
+		Timeout: s.timeout,
```

Comment:
Should disable HTTP compression if partial-chosen-plaintext attacks are a possibility, a la [BREACH](https://en.wikipedia.org/wiki/BREACH).  Probably also worth looking at any other places we use HTTP over TLS (or other protocols) and see if they have compression on too.

Reply:
+1

Reply:
It looks like there aren't any other places.